_posts/2020-04-08-tf-federated-intro/tf-federated-intro.Rmd

```diff
@@ -0,0 +1,464 @@
+
+```{r}
+state_and_metrics <- iterative_process$`next`(state, federated_train_data)
+
+state <- state_and_metrics[0]
+state
+```
+
+    <model=<trainable=<[[ 9.9695253e-06 -8.5083229e-05 -8.9266898e-05 ... -7.7834651e-05
+      -9.4819807e-05  3.4227365e-04]
+     [-5.4778640e-05 -1.5390900e-04 -1.7912561e-04 ... -1.4122366e-04
+      -2.4614178e-04  7.7663612e-04]
+     [-1.9177950e-04 -9.0706220e-05 -2.9841764e-04 ... -2.2249141e-04
+      -4.1685964e-04  1.1348884e-03]
+     ...
+     [-1.3832574e-03 -5.3664664e-04 -3.6622395e-04 ... -9.0854493e-04
+       4.9618416e-04  2.6899918e-03]
+     [-7.7253254e-04 -2.4583895e-04 -8.3220737e-05 ... -4.5274393e-04
+       2.6396243e-04  1.7454443e-03]
+     [-2.4157032e-04 -1.3836231e-05  5.0371520e-05 ... -1.0652864e-04
+       1.5947431e-04  4.5250656e-04]],[-0.01264258  0.00974309  0.00814162  0.00846065 -0.0162328   0.01627758
+     -0.00445857 -0.01607843  0.00563046  0.00115899]>,non_trainable=<>>,optimizer_state=<1>,delta_aggregate_state=<>,model_broadcast_state=<>>
+
+```{r}
+metrics <- state_and_metrics[1]
+metrics
+```
+
+    <sparse_categorical_accuracy=0.5710999965667725,loss=1.8662642240524292,keras_training_time_client_sum_sec=0.0>
+
+Let's train for a few more epochs, keeping track of accuracy:
+
+```{r}
+num_rounds <- 20
+
+for (round_num in (2:num_rounds)) {
+  state_and_metrics <- iterative_process$`next`(state, federated_train_data)
+  state <- state_and_metrics[0]
+  metrics <- state_and_metrics[1]
+  cat("round: ", round_num, "  accuracy: ", round(metrics$sparse_categorical_accuracy, 4), "\n")
+}
+```
+
+    round:  2   accuracy:  0.6949 
+    round:  3   accuracy:  0.7132 
+    round:  4   accuracy:  0.7231 
+    round:  5   accuracy:  0.7319 
+    round:  6   accuracy:  0.7404 
+    round:  7   accuracy:  0.7484 
+    round:  8   accuracy:  0.7557 
+    round:  9   accuracy:  0.7617 
+    round:  10   accuracy:  0.7661 
+    round:  11   accuracy:  0.7695 
+    round:  12   accuracy:  0.7728 
+    round:  13   accuracy:  0.7764 
+    round:  14   accuracy:  0.7788 
+    round:  15   accuracy:  0.7814 
+    round:  16   accuracy:  0.7836 
+    round:  17   accuracy:  0.7855 
+    round:  18   accuracy:  0.7872 
+    round:  19   accuracy:  0.7885 
+    round:  20   accuracy:  0.7902 
+
+Training accuracy is increasing continuously. These values represent averages of
+*local* accuracy measurements, so in the real world, might well be overly
+optimistic (with each client overfitting on their respective data). So
+supplementing federated training, a federated evaluation process would need to
+be built in order to get a realistic view on performance. This is a topic to
+come back to when more related TFF documentation is available.
+
+### Conclusion
+
+We hope you've enjoyed this first introduction to TFF using R. Certainly at this
+time, it is too early for productive use; and for application in research (e.g., adversarial attacks on federated learning)
```

Comment:
By productive use, do you mean use in production? If yes, then I'd say use in production, but, if no, you're good. 🙂

Reply:
yeah, I do... wow, here I really would have thought that works, but doesn't seem like...